### PR TITLE
PIM-3881: add a job launcher

### PR DESCRIPTION
**Pull request by skeleton**

### CODE REVIEW

File: Manager/JobLauncher.php [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    public function __construct(
+        ObjectManager $objectManager,
+        EventDispatcherInterface $eventDispatcher,
+        $jobExecutionClass,
+        $rootDir,
+        $environment
+    ) {
+        $this->objectManager      = $objectManager;
+        $this->eventDispatcher    = $eventDispatcher;
+        $this->jobExecutionClass  = $jobExecutionClass;
```
Comment (solivier) [inline Manager/JobLauncher.php]: extra spaces

File: Manager/JobLauncher.php [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+use Symfony\Component\EventDispatcher\GenericEvent;
+use Symfony\Component\Process\PhpExecutableFinder;
+use Symfony\Component\Security\Core\User\UserInterface;
+
+/**
+ * @author    Marie Bochu <marie.bochu@akeneo.com>
+ * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/MIT MIT
+ */
+class JobLauncher
```
Comment (grena) [inline Manager/JobLauncher.php]: I think this launcher will be called `SimpleJobLauncher`, to confirm with @nidup and @BitOne.

File: Manager/JobLauncher.php [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        EventDispatcherInterface $eventDispatcher,
+        $jobExecutionClass,
+        $rootDir,
+        $environment
+    ) {
+        $this->objectManager      = $objectManager;
+        $this->eventDispatcher    = $eventDispatcher;
+        $this->jobExecutionClass  = $jobExecutionClass;
+        $this->rootDir            = $rootDir;
+        $this->environment        = $environment;
```
Comment (grena) [inline Manager/JobLauncher.php]: There is 1 extra space before each `=`

File: Manager/JobLauncher.php [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    /**
+     * Has config
+     *
+     * @param string $config
+     *
+     * @return bool
+     */
+    public function hasConfig($config)
```
Comment (grena) [inline Manager/JobLauncher.php]: `hasConfigKey`, `hasConfigIndex`  ?

File: Manager/JobLauncher.php [hunk 5]
```diff
@@ -0,0 +1,149 @@
+<?php
+
+namespace Akeneo\Bundle\BatchBundle\Manager;
```
Comment (BitOne) [inline Manager/JobLauncher.php]: Manager is not precise enough. I would say Launcher as namespace ;)

File: Manager/JobLauncher.php [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+use Akeneo\Bundle\BatchBundle\Entity\JobExecution;
+use Akeneo\Bundle\BatchBundle\Entity\JobInstance;
+use Doctrine\Common\Persistence\ObjectManager;
+use Pim\Bundle\ImportExportBundle\Event\JobProfileEvents;
```
Comment (nidup) [inline Manager/JobLauncher.php]: we should not depends on the PIM here, we should move events

File: Manager/JobLauncher.php (hunk 2 above)
Comment (BitOne) [inline Manager/JobLauncher.php]: Yes, agree. With a nice interface to implement new Launcher.

File: Manager/JobLauncher.php [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    /**
+     * Instantiate a new job execution
+     *
+     * @param JobInstance   $jobInstance
+     * @param UserInterface $user
+     *
+     * @return JobExecution
+     * @throws \Exception
+     */
+    protected function create(JobInstance $jobInstance, UserInterface $user)
```
Comment (BitOne) [inline Manager/JobLauncher.php]: You should use the existing method to do that:
https://github.com/akeneo/BatchBundle/blob/master/Job/DoctrineJobRepository.php#L68

File: Manager/JobLauncher.php [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+
+/**
+ * @author    Marie Bochu <marie.bochu@akeneo.com>
+ * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/MIT MIT
+ */
+class JobLauncher
+{
+    /** @var ObjectManager */
+    protected $objectManager;
```
Comment (BitOne) [inline Manager/JobLauncher.php]: If the ObjectManager is there to manage Job entities, you should inject the JobRepository, as job can be stored outside of a DB for example, it depends of the concrete implementation of the JobRepositoryInterface.
See https://github.com/akeneo/BatchBundle/blob/master/Resources/config/services.yml#L23

File: Manager/JobLauncher.php [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+            $executionId,
+            !empty($rawConfiguration) ? sprintf('--config="%s"', $rawConfiguration) : '',
+            $this->rootDir
+        );
+
+        // Please note we do not use Symfony Process as it has some problem
+        // when executed from HTTP request that stop fast (race condition that makes
+        // the process cloning fail when the parent process, i.e. HTTP request, stops
+        // at the same time)
+        exec($cmd . ' &');
```
Comment (BitOne) [inline Manager/JobLauncher.php]: Would be better to handle the command to another method, like launchInBackground() in order to allow easy override of the call (the '&' works only in Unix env) and better placement of comment.

File: Manager/JobLauncher.php [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+    /**
+     * Has config
+     *
+     * @param string $config
+     *
+     * @return bool
+     */
+    public function hasConfig($config)
+    {
+        return isset($this->config[$config]) && true === $this->config[$config];
```
Comment (BitOne) [inline Manager/JobLauncher.php]: So it means that $this->config is an array containing only true or false ?

Because on this test, $this->config[$config] must contains extactly true

File: Manager/JobLauncher.php (hunk 7 above)
Comment (skeleton) [inline Manager/JobLauncher.php]: Sure ? 
Actually, class name is in yml configuration file. If I use DoctrineJobRepository.php::create, class is hard coded :(

Comment (BitOne) [inline Manager/JobLauncher.php]: Yes, indeed. But the problem is in DoctrineJobRepository, it should not be circumvent here. It's not the reponsability of the Launcher to manage the persistence topic.

Comment (BitOne) [inline Manager/JobLauncher.php]: And we could do another PR to fix the DoctrineJobRepository problem.

File: Manager/JobLauncher.php (hunk 10 above)
Comment (skeleton) [inline Manager/JobLauncher.php]: Yes, there is only boolean to allowed some options

Comment (BitOne) [inline Manager/JobLauncher.php]: So the naming of the method is not right, because it check if the config is set to true not if it's there.

File: Launcher/SimpleJobLauncher.php [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+    /**
+     * Constructor
+     *
+     * @param JobRepositoryInterface   $jobRepository
+     * @param string                   $rootDir
+     * @param string                   $environment
+     */
+    public function __construct(JobRepositoryInterface $jobRepository, $rootDir, $environment)
+    {
+        $this->jobRepository     = $jobRepository;
```
Comment (jjanvier) [inline Launcher/SimpleJobLauncher.php]: align

Comment (BitOne) [inline Launcher/SimpleJobLauncher.php]: Maybe too many spaces ?

File: Launcher/SimpleJobLauncher.php [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+     * {@inheritdoc}
+     */
+    public function launch(JobInstance $jobInstance, UserInterface $user, $rawConfiguration = null)
+    {
+        $jobExecution = $this->create($jobInstance, $user);
+        $executionId  = $jobExecution->getId();
+        $pathFinder   = new PhpExecutableFinder();
+
+        if ($this->isConfigTrue('upload')) {
+            $rawConfiguration.= ' ' . addslashes(json_encode($jobInstance->getJob()->getConfiguration()));
```
Comment (jjanvier) [inline Launcher/SimpleJobLauncher.php]: $rawConfigurationSPACE.=

File: Launcher/SimpleJobLauncher.php [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+            $this->rootDir
+        );
+
+        $this->launchInBackground($cmd);
+
+        return $jobExecution;
+    }
+
+    /**
+     * {@inheritdoc}
```
Comment (jjanvier) [inline Launcher/SimpleJobLauncher.php]: nope

File: Launcher/SimpleJobLauncher.php [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+     */
+    public function launchInBackground($cmd)
```
Comment (jjanvier) [inline Launcher/SimpleJobLauncher.php]: why this method exists ? and why public ?

File: Launcher/SimpleJobLauncher.php [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    /**
+     * Is key is true in configuration
+     *
+     * @param string $key
+     *
+     * @return bool
+     */
+    public function isConfigTrue($key)
```
Comment (jjanvier) [inline Launcher/SimpleJobLauncher.php]: why public ?

File: Launcher/SimpleJobLauncher.php [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function launchInBackground($cmd)
+    {
+        // Please note we do not use Symfony Process as it has some problem
+        // when executed from HTTP request that stop fast (race condition that makes
+        // the process cloning fail when the parent process, i.e. HTTP request, stops
+        // at the same time)
```
Comment (BitOne) [inline Launcher/SimpleJobLauncher.php]: I would put this comment in the method phpDoc ;) Not in the interface, but after the {@inheritdoc}

File: Launcher/SimpleJobLauncher.php [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    /**
+     * Set config
+     *
+     * @param array $config
+     *
+     * @return JobLauncher
+     */
+    public function setConfig(array $config = [])
```
Comment (jjanvier) [inline Launcher/SimpleJobLauncher.php]: no need in the interface ?

File: Launcher/SimpleJobLauncher.php [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+    {
+        $this->config = $config;
+
+        return $this;
+    }
+
+    /**
+     * Is key is true in configuration
```
Comment (BitOne) [inline Launcher/SimpleJobLauncher.php]: Is key true

File: Launcher/SimpleJobLauncher.php (hunk 17 above)
Comment (jjanvier) [inline Launcher/SimpleJobLauncher.php]: no need of `= []` in a setter

File: Launcher/JobLauncherInterface.php [hunk 19]
```diff
@@ -0,0 +1,25 @@
+<?php
+
+namespace Akeneo\Bundle\BatchBundle\Launcher;
+
+use Akeneo\Bundle\BatchBundle\Entity\JobInstance;
+use Symfony\Component\Security\Core\User\UserInterface;
+
+/**
```
Comment (jjanvier) [inline Launcher/JobLauncherInterface.php]: phpdoc

File: Launcher/SimpleJobLauncher.php (hunk 14 above)
Comment (skeleton) [inline Launcher/SimpleJobLauncher.php]: See @BitOne comment: https://github.com/akeneo/BatchBundle/pull/61#discussion_r28506928

File: Launcher/SimpleJobLauncher.php [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+
+    /**
+     * Create a jobExecution
+     *
+     * @param JobInstance   $jobInstance
+     * @param UserInterface $user
+     *
+     * @return JobExecution
+     */
+    protected function create(JobInstance $jobInstance, UserInterface $user)
```
Comment (BitOne) [inline Launcher/SimpleJobLauncher.php]: Maybe createJobExecution.

File: Launcher/SimpleJobLauncher.php (hunk 14 above)
Comment (BitOne) [inline Launcher/SimpleJobLauncher.php]: And yes, no need to put that in the interface or make it public. It's all about specific implementation.

**Comment by jjanvier** [comment]
GTM
